[Lai Yuen Sheng] iP

seanlaiys opened the pull request:
# Alfred Chatbot
> "Why Do We Fall, Sir? So That We Can Learn To Pick Ourselves Up." – Alfred Pennyworth ([Source](https://www.cbr.com/batman-dark-knight-trilogy-alfred-best-quotes/))

Alfred is your trusty butler to count on. He is,
- Approachable
- Easy to talk to
- ~~Helpful~~ SUPER helpful

All you need to do is,
1. Download from [here](https://github.com/seanlaiys/ip/releases/tag/A-JAR) 
2. Double-Click it 
3. Add you tasks and let your friendly butler manage them :smiley:

And Alfred's services are **FREE**!

Features:
- [X] Managing deadlines
- [X] Managing events
- [X] Managing tasks
- [X] Setting task priority

Commit:
Add Gradle support

seanlaiys committed:
Created Chatbot Alfred

seanlaiys committed:
Added methods to greet user and read command

seanlaiys committed:
Created class to read commands bye, list and blah.

seanlaiys committed:
Created list of inputable commands.

seanlaiys committed:
Created methods that print messages depending on user input.

seanlaiys committed:
Added Storage class.

seanlaiys committed:
Added documentation.

seanlaiys committed:
Enabled Adding of Tasks.

seanlaiys committed:
Created new Task Class.

seanlaiys committed:
Enabled marking and unmarking of tasks.

seanlaiys committed:
Added todo, event and deadline.

seanlaiys committed:
Ran Tests.

seanlaiys committed:
Handle Errors.

seanlaiys committed:
Added delete method.

seanlaiys committed:
Change name to Parser

seanlaiys committed:
Change CommandReader class name to Parser

seanlaiys committed:
Change documentation

seanlaiys committed:
Change Storage class and add TaskList class

seanlaiys committed:
Change Storage class

seanlaiys committed:
Change Task class to abstract

seanlaiys committed:
Add Deadline class

seanlaiys committed:
Add ToDo class

seanlaiys committed:
Add taskFile for saving tasks

seanlaiys committed:
Add Event class

seanlaiys committed:
Change CommandReader name to Parser

seanlaiys committed:
Add TaskList class

seanlaiys committed:
Update data

seanlaiys committed:
Add LocalDate and LocalTime

seanlaiys committed:
Add LocalDate and LocalTime

seanlaiys committed:
Add LocalDate and LocalTime

seanlaiys committed:
Add Invalid Date message

seanlaiys committed:
Add LocalDate and LocalTime input

seanlaiys committed:
Add LocalDate and LocalTime

seanlaiys committed:
Add abstract method getDate()

seanlaiys committed:
Add getTasksByDate method

seanlaiys committed:
Add getDate method

seanlaiys committed:
Update TaskFile

seanlaiys committed:
Merge branch 'master' into branch-Level-7

seanlaiys committed:
Merge branch 'master' into branch-Level-7

seanlaiys committed:
Add ToDate

seanlaiys committed:
Merge branch 'master' into branch-Level-8

seanlaiys committed:
Update Documentation

seanlaiys committed:
Add More OOP

seanlaiys committed:
Organize into Packages

seanlaiys committed:
Add JUnit Tests

seanlaiys committed:
Create a JAR File

seanlaiys committed:
Add JavaDoc comments

seanlaiys committed:
Follow the Coding Standard

seanlaiys committed:
Revert "Follow the Coding Standard"

This reverts commit 62c468281e08b6f0d860d4b4230edf7c872212ce.

seanlaiys committed:
Follow the Coding Standard

seanlaiys committed:
Merge branch 'master' of https://github.com/seanlaiys/ip into branch-A-JavaDoc

seanlaiys committed:
Follow Coding Standards

seanlaiys committed:
Merge branch 'master' into branch-A-CodingStandard

seanlaiys committed:
Follow Coding Standard

seanlaiys committed:
Merge branch 'branch-A-CodingStandard' of https://github.com/seanlaiys/ip into branch-A-CodingStandard

seanlaiys committed:
Edit JavaDoc

seanlaiys committed:
Merge branch 'master' into branch-A-JavaDoc

seanlaiys committed:
Merge branch 'master' of https://github.com/seanlaiys/ip

seanlaiys committed:
Merge branch 'branch-A-CodingStandard' of https://github.com/seanlaiys/ip into branch-A-CodingStandard

seanlaiys committed:
Edit to follow Coding Standard

seanlaiys committed:
Add Find Feature

whtan99 commented on src/main/java/alfred/action/Parser.java:
```diff
@@ -0,0 +1,246 @@
+                }
+            } else if (input.contains(Commands.COMMAND_EVENT)) {
+                String[] descriptions = input.split(Commands.COMMAND_EVENT);
+                if (descriptions.length == 0) {
+                    throw new MissingDescriptionException();
+                } else {
+                    String[] inputs = input.split("/at ");
+                    if (!isValidTime(inputs[1]) || !isValidDate(inputs[1])) {
+                        throw new InvalidDateException();
+                    } else {
+                        try {
+                            Task newTask = new Event(inputs[0], getDate(inputs[1]),
+                                    getTime(inputs[1]));
+                            currentTasks.addTasks(newTask);
+                            input = Ui.sayAdd(user_input, newTask, currentTasks);
+                            currentStorage.appendTaskToFile(newTask);
+                        } catch (IOException e) {
+                            e.printStackTrace();
+                        }
+                    }
+                }
+            } else if (input.contains(Commands.COMMAND_DELETE)) {
+                String[] descriptions = input.split(Commands.COMMAND_DELETE);
+                if (descriptions.length == 0) {
+                    throw new EmptyInputException();
+                } else {
+                    int taskNumber = findDigit(input);
+                    Task removedTask = currentTasks.getTask(taskNumber);
+                    currentTasks.removeTasks(taskNumber);
+                    input = Ui.sayDelete(user_input, removedTask, currentTasks);
+                    try {
+                        currentStorage.writeTasksToFile(currentTasks);
+                    } catch (IOException e) {
+                        e.printStackTrace();
+                    }
+                }
+            } else if (input.equals("")) {
+                throw new EmptyInputException();
+            } else {
+                throw new InvalidInputException();
```
Would this be less troublesome to maintain if you can use case/switch statement instead?

whtan99 commented on src/main/java/alfred/action/Storage.java:
```diff
@@ -0,0 +1,158 @@
+/**
+ *
+ * Stores the current tasks in an <code>ArrayList</code>.
+ *
```
Would deleting this empty line be more synchronized with the coding standard?

whtan99 commented on src/main/java/alfred/action/TaskList.java:
```diff
@@ -0,0 +1,132 @@
+     * the method, which will be stored in tasks.
+     *
+     * @param newTask the new Task object to be added
+     *
```
Similar to above, perhaps removing this extra line for the java doc would be more synchronised with the coding standards?

whtan99 commented on src/main/java/alfred/task/Task.java:
```diff
@@ -0,0 +1,64 @@
+     * @param keyword keyword input by user
+     * @return a boolean of whether the task contains the keyword
+     */
+    public boolean contains(String keyword) {
+        if (task.contains(keyword)) {
+            return true;
+        }
+        return false;
+    }
```
```suggestion
return task.contains(keyword)
```

whtan99 replied:
Would writing the code in this format be more concise?

whtan99 commented on src/main/java/alfred/task/ToDo.java:
```diff
@@ -0,0 +1,49 @@
+
+import java.time.LocalDate;
+/**
+ * Encapsulates a <code>ToDo</code> which inherits from <code>Task</code>.
```
Could I clarify what `<code>` header is for? On the pull request it just seems like there might be some formatting errors. Thanks!

whtan99 commented on src/test/java/alfredtest/TaskListTest.java:
```diff
@@ -0,0 +1,36 @@
+public class TaskListTest {
+
+    @Test
+    public void testAddTaskToTaskList() {
```
```suggestion
    public void addTaskToTaskListTest() {
```

whtan99 replied:
Would putting the Test at the end of the method be clearer for the user?

seanlaiys replied:
Thanks! Will make the necessary changes.

whtan99 commented on src/main/java/alfred/action/Ui.java:
```diff
@@ -0,0 +1,154 @@
+ *
+ */
+public class Ui {
+    private static final String LOGO =
+            "    __      \n"
+            + "   /  \\    \n"
+            + "  / /\\ \\  \n"
+            + " / /__\\ \\  \n"
+            + "/_/    \\_\\ \n";
+
+    private static final String GREETING =
+            "Greetings, my name is Alfred. \n" + "How may I assist you?";
+
+    private static final String LIST = "Here is your list.";
+
+    private static final String BLAH = "Blah.";
+
+    private static final String BYE = "It has been great pleasure to be at your service. Goodbye.";
+
+    private static final String ADD = "Got it. I've added this task:";
+
+    private static final String DELETE = "Got it. I've removed this task:";
+
+    private static final String MARK = "You have successfully completed the following task:";
+
+    private static final String UNMARK = "You requested to mark the following task as undone:";
+
+    private static final String TASK_SIZE = "Here is the number of tasks in the list:";
+
+    private static final String WARNING = "Do be careful with your input or you have to re-run me again";
```
I really like how the bot's reply are added into class variables, which makes it easier for the user to read to pre-determined replies of the bot instead of going into each method!

whtan99 reviewed (commented):
Looks good 👍 
Some minor edits to java docs requested.

seanlaiys committed:
Merge branch 'master' into add-gradle-support

seanlaiys committed:
Add Checkstyle

seanlaiys committed:
Add GUI

eman-kom commented on src/main/java/alfred/action/Parser.java:
```diff
@@ -0,0 +1,253 @@
+import alfred.task.Event;
+import alfred.task.Task;
+import alfred.task.ToDo;
+/**
```
There should be a space before the javadoc separating the import from the javadoc

seanlaiys replied:
Thanks! Will make the necessary changes.

eman-kom commented on src/main/java/alfred/task/Task.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    /**
+     * Marks a Tas object undone.
```
Task mispelled as Tas

eman-kom commented on src/main/java/alfred/exception/EmptyInputException.java:
```diff
@@ -0,0 +1,17 @@
+     *
+     */
+    public EmptyInputException() {
+        super(NO_DESCRIPTION);
```
Would it be possible to move the string into the super()
```suggestion
        super("Please try again with a description in mind!");
```

eman-kom commented on src/main/java/alfred/exception/MissingDescriptionException.java:
```diff
@@ -0,0 +1,16 @@
+     * Creates a new exception object for missing description in tasks.
+     */
+    public MissingDescriptionException() {
+        super(EMPTY);
```
Would it be possible to move the string into the super()
```suggestion
        super("I can't help you if you don't type anything");
```

eman-kom commented on src/main/java/alfred/exception/InvalidInputException.java:
```diff
@@ -0,0 +1,17 @@
+     * Creates a new exception object for invalid input.
+     */
+    public InvalidInputException() {
+        super(UNSURE);
```
Would it be possible to move the string into the super()
```suggestion
        super("OOPS!!! I'm sorry, "
            + "but I don't know what that means :-(");
```

eman-kom commented on src/main/java/alfred/exception/InvalidDateException.java:
```diff
@@ -0,0 +1,16 @@
+     * Creates a new exception object for invalid date.
+     */
+    public InvalidDateException() {
+        super(INVALID_DATE);
```
Would it be possible to move the string into the super()
```suggestion
        super("Please try again with a valid date and time DD/MM/YYYY HHMM");
```

eman-kom commented on src/main/java/alfred/action/Storage.java:
```diff
@@ -0,0 +1,158 @@
+        while (s.hasNext()) {
+            current = s.nextLine();
+            if (current.contains(Commands.COMMAND_TODO)) {
+                String[] descriptions = current.split(Commands.COMMAND_TODO);
+                Task newTask = new ToDo(Commands.COMMAND_TODO + descriptions[1],
+                        current.contains("[X]"));
+                tasks.addTasks(newTask);
+            } else if (current.contains(Commands.COMMAND_EVENT)) {
+                String[] descriptions = current.split(Commands.COMMAND_EVENT);
+                String openBracket = "\\(";
+                String closeBracket = "\\)";
+                String[] newDescriptions = descriptions[1].split(openBracket);
+                String date = newDescriptions[1].split(closeBracket)[0];
+                String[] dateSplits = date.split("at: ")[1].split(" ");
+                date = dateSplits[0] + " " + dateSplits[1] + " " + dateSplits[2];
+                String time = dateSplits[3] + dateSplits[4];
+                DateTimeFormatter formatterDate = DateTimeFormatter.ofPattern("MMM d yyyy");
+                DateTimeFormatter formatterTime = DateTimeFormatter.ofPattern("hh:mma");
+                LocalDate newDate = LocalDate.parse(date, formatterDate);
+                LocalTime newTime = LocalTime.parse(time, formatterTime);
+                Task newTask = new Event(Commands.COMMAND_EVENT + newDescriptions[0],
+                        newDate, newTime, current.contains("[X]"));
+                tasks.addTasks(newTask);
+            } else {
+                String[] descriptions = current.split(Commands.COMMAND_DEADLINE);
+                String openBracket = "\\(";
+                String closeBracket = "\\)";
+                String[] newDescriptions = descriptions[1].split(openBracket);
+                String date = newDescriptions[1].split(closeBracket)[0];
+                String[] dateSplits = date.split("by: ")[1].split(" ");
+                date = dateSplits[0] + " " + dateSplits[1] + " " + dateSplits[2];
+                String time = dateSplits[3] + dateSplits[4];
+                DateTimeFormatter formatterDate = DateTimeFormatter.ofPattern("MMM d yyyy");
+                DateTimeFormatter formatterTime = DateTimeFormatter.ofPattern("hh:mma");
+                LocalDate newDate = LocalDate.parse(date, formatterDate);
+                LocalTime newTime = LocalTime.parse(time, formatterTime);
+                Task newTask = new Deadline(Commands.COMMAND_DEADLINE + newDescriptions[0],
+                        newDate, newTime, current.contains("[X]"));
+                tasks.addTasks(newTask);
+            }
```
Could there be line breaks to breakup the chunk of text to make it more readable?

eman-kom reviewed (commented):
Nice work 👍 
Keep it up!

seanlaiys commented:
> Looks good 👍 Some minor edits to java docs requested.

Thanks 👍 ! will make the necessary changes.

seanlaiys committed:
Update gradle to enable assertions

Assertions are used to define assumptions about the program state
so that the runtime can verify them.

An assertion failure indicates a possible bug in the code, because the
code has resulted in a program state that violates an assumption about
how the code should behave.

seanlaiys committed:
Add assertions

An assertion failure indicates a possible bug in the code, because the
code has resulted in a program state that violates an assumption about
how the code should behave.

As such, adding assertions ensures important variables in the selected
files do not contain null values or indexes out of the bounds of a list.

seanlaiys committed:
Enhance GUI appearance

It is essential to ensure the appearance of the GUI is optimised and
visually aesthetic, to enhance user's experience. This is because
this is the platform where users interact with the application.

The theme of the chat is greatly tied with Alfred's background and the
client he serves.

seanlaiys committed:
Merge pull request #1 from seanlaiys/branch-A-Assertions

Add assertions to document important assumptions

seanlaiys committed:
Improve Code Quality

Code quality is an important aspect to ensure readability of code. This
helps oneself in recognising and understanding one's code and at the
same time, aid others in comprehending one's code.

By reducing unnecessarily long methods and deep nesting, codes can be
better read and understood.

seanlaiys committed:
Merge pull request #2 from seanlaiys/branch-A-CodeQuality

Improve Code Quality

seanlaiys committed:
Add Priority

By adding the priority attribute to Task, Task can implement
Comparable<Task> and their priority can be compared, which enables
features to be added such as the use of priority queue to generate
list of tasks in order of priority.

Users can now view and update the priority of their tasks and decide
what is important at hand.

seanlaiys committed:
Improve Coding Standard

seanlaiys committed:
Improve GUI

seanlaiys committed:
Add UI screenshot

seanlaiys committed:
Remove bug

seanlaiys committed:
Add user guide

seanlaiys committed:
Update GUI screenshot

seanlaiys committed:
Update GUI screenshot

seanlaiys committed:
Add stage name

seanlaiys committed:
Set theme time machine

seanlaiys committed:
Move config file to docs

seanlaiys committed:
Set theme jekyll-theme-time-machine

seanlaiys committed:
Remove bugs and Shorten Long Methods

While long methods should often be avoided, the justification for
exceeding the usual limit of 30 lines for the method parseCommand() is
due to a significant number of commands a user can execute. The method
has been shortened from 170 lines to 43 lines, by using SLAP.

This update has also removed the bugs associated with date and time,
which also affects the storage of the tasks when tasks with invalid
dates pass through. Users are highly discouraged from modifying data
directly from the file as that affects the parsing of tasks from storage
to the TaskList. However, in the event of such modifications, an
exception will occur and users will be advised to delete data from their
file to make sure the application works properly.

seanlaiys committed:
Merge pull request #5 from seanlaiys/branch-update-release

Remove bugs and Shorten Long Methods

seanlaiys committed:
Update User Guide and Test Cases.

seanlaiys committed:
Merge pull request #6 from seanlaiys/branch-update-release

Update User Guide and Test Cases.

seanlaiys committed:
Edit Wrong Command Exit to Bye

seanlaiys committed:
Merge pull request #7 from seanlaiys/branch-update-release

Edit Wrong Command Exit to Bye

seanlaiys committed:
Remove instances of Exit Command and Change to Bye

seanlaiys committed:
Merge pull request #8 from seanlaiys/branch-update-release

Remove instances of Exit Command and Change to Bye

seanlaiys committed:
Update README.md
